index.js

```diff
@@ -374,30 +375,73 @@ Pbf.prototype = {
     }
 };
 
-function readVarintRemainder(val, pbf) {
-    var buf = pbf.buf, b;
+function readVarintRemainder(l, u, p) {
+    var buf = p.buf,
+        h, b;
 
-    b = buf[pbf.pos++]; val += (b & 0x7f) * 0x10000000;         if (b < 0x80) return val;
-    b = buf[pbf.pos++]; val += (b & 0x7f) * 0x800000000;        if (b < 0x80) return val;
-    b = buf[pbf.pos++]; val += (b & 0x7f) * 0x40000000000;      if (b < 0x80) return val;
-    b = buf[pbf.pos++]; val += (b & 0x7f) * 0x2000000000000;    if (b < 0x80) return val;
-    b = buf[pbf.pos++]; val += (b & 0x7f) * 0x100000000000000;  if (b < 0x80) return val;
-    b = buf[pbf.pos++]; val += (b & 0x7f) * 0x8000000000000000; if (b < 0x80) return val;
+    b = buf[p.pos++]; h  = (b & 0x70) >> 4;  if (b < 0x80) return toNum(l, h, u);
+    b = buf[p.pos++]; h |= (b & 0x7f) << 3;  if (b < 0x80) return toNum(l, h, u);
+    b = buf[p.pos++]; h |= (b & 0x7f) << 10; if (b < 0x80) return toNum(l, h, u);
+    b = buf[p.pos++]; h |= (b & 0x7f) << 17; if (b < 0x80) return toNum(l, h, u);
+    b = buf[p.pos++]; h |= (b & 0x7f) << 24; if (b < 0x80) return toNum(l, h, u);
+    b = buf[p.pos++]; h |= (b & 0x01) << 31; if (b < 0x80) return toNum(l, h, u);
 
     throw new Error('Expected varint not more than 10 bytes');
 }
 
+function toNum(low, high, isUnsigned) {
+    if (isUnsigned) {
+        return ((high >>> 0) * 0x100000000) + (low >>> 0);
+    }
+
+    return high * 0x100000000 + (low >>> 0);
+}
+
 function writeBigVarint(val, pbf) {
+    var low, high;
+
+    if (val >= 0) {
+        low  = (val % 0x100000000) | 0;
+        high = (val / 0x100000000) | 0;
+    } else {
```

Comment:
I'm not a fan of the crazy twos-compliment math going on in this `else` statement in order to get the sign bit properly on a negative number and the 32nd bit the low bits. However, I can't think of a another way to accomplish it. It would be fantastic if you are able to think of another way.